src/app/InteractionModelEngine.h

```diff
@@ -0,0 +1,182 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines objects for a CHIP Interaction Data model Engine which handle unsolicitied IM message, and
+ *      manage different kinds of IM client and handlers.
+ *
+ */
+
+#pragma once
+
+#ifndef _CHIP_INTERACTION_MODEL_ENGINE_H
+#define _CHIP_INTERACTION_MODEL_ENGINE_H
+
+#include <app/MessageDef.h>
+#include <core/CHIPCore.h>
+#include <map>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+#include <app/Command.h>
+#include <app/CommandHandler.h>
+#include <app/CommandSender.h>
+
+#define CHIP_MAX_NUM_COMMAND_HANDLER_OBJECTS 1
+#define CHIP_MAX_NUM_COMMAND_SENDER_OBJECTS 1
+
+namespace chip {
+namespace app {
+
+typedef void (*CommandCbFunct)(chip::TLV::TLVReader & aReader, Command * apCommandObj);
+
+/**
+ * @class InteractionModelEngine
+ *
+ * @brief This is a singleton hosting all CHIP IM unsolicitted message and manage IM related clients and handlers.
+ *
+ */
+class InteractionModelEngine : public Messaging::ExchangeDelegate
+{
+public:
+    enum EventID
+    {
+        kEvent_OnIncomingInvokeCommandRequest =
+            0, ///< Called when an incoming invoke command request has arrived before applying commands..
+    };
+
+    /**
+     * Incoming parameters sent with events generated directly from this component
+     *
+     */
+    union InEventParam
+    {
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+        struct
+        {
+            const PacketHeader * mpPacketHeader; ///< A pointer to the message information for the request
+        } mIncomingInvokeCommandRequest;
+    };
+
+    /**
+     * Outgoing parameters sent with events generated directly from this component
+     *
+     */
+    union OutEventParam
+    {
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+
+        struct
+        {
+            bool mShouldContinueProcessing; ///< Set to true if update is allowed.
+        } mIncomingInvokeCommandRequest;
+    };
+
+    /**
+     * @brief Set the event back function and pointer to associated state object for SubscriptionEngine specific call backs
+     *
+     * @param[in]  aAppState    A pointer to application layer supplied state object
+     * @param[in]  aEvent       A function pointer for event call back
+     * @param[in]  aInParam     A const reference to the input parameter for this event
+     * @param[out] aOutParam    A reference to the output parameter for this event
+     */
+    typedef void (*EventCallback)(void * const aAppState, EventID aEvent, const InEventParam & aInParam, OutEventParam & aOutParam);
```

Comment:
nit: Remove `const` qualifier from `aAppState` as it has no effect.

and two below (aAppState, EventCallback).

Reply:
agree, thank you.